Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+PathComponent.swift

```diff
@@ -37,88 +37,444 @@ extension PathHierarchy {
         
-        func isValidHash(_ hash: String) -> Bool {
+        func isValidHash(_ hash: Substring) -> Bool {
             var index: UInt8 = 0
             for char in hash.utf8 {
                 guard index <= 5, (48...57).contains(char) || (97...122).contains(char) else { return false }
                 index += 1
             }
-            return true
+            return index > 0
         }
         
-        if knownSymbolKinds.contains(hash) {
+        if knownSymbolKinds.contains(String(hash)) {
             // The parsed hash value is a symbol kind
             return PathComponent(full: full, name: name, kind: hash, hash: nil)
         }
         if let languagePrefix = knownLanguagePrefixes.first(where: { hash.starts(with: $0) }) {
             // The hash is actually a symbol kind with a language prefix
-            return PathComponent(full: full, name: name, kind: String(hash.dropFirst(languagePrefix.count)), hash: nil)
+            return PathComponent(full: full, name: name, kind: hash.dropFirst(languagePrefix.count), hash: nil)
         }
         if !isValidHash(hash) {
             // The parsed hash is neither a symbol not a valid hash. It's probably a hyphen-separated name.
-            return PathComponent(full: full, name: full, kind: nil, hash: nil)
+            return PathComponent(full: full, name: full[...], kind: nil, hash: nil)
         }
         
         if let dashIndex = name.lastIndex(of: "-") {
-            let kind = String(name[dashIndex...].dropFirst())
-            let name = String(name[..<dashIndex])
-            if knownSymbolKinds.contains(kind) {
+            let kind = name[dashIndex...].dropFirst()
+            let name = name[..<dashIndex]
+            if knownSymbolKinds.contains(String(kind)) {
                 return PathComponent(full: full, name: name, kind: kind, hash: hash)
             } else if let languagePrefix = knownLanguagePrefixes.first(where: { kind.starts(with: $0) }) {
-                let kindWithoutLanguage = String(kind.dropFirst(languagePrefix.count))
+                let kindWithoutLanguage = kind.dropFirst(languagePrefix.count)
                 return PathComponent(full: full, name: name, kind: kindWithoutLanguage, hash: hash)
             }
         }
         return PathComponent(full: full, name: name, kind: nil, hash: hash)
     }
     
-    static func joined<PathComponents>(_ pathComponents: PathComponents) -> String where PathComponents: Sequence, PathComponents.Element == PathComponent {
-        return pathComponents.map(\.full).joined(separator: "/")
+    static func split(_ path: String) -> (componentSubstrings: [Substring], isAbsolute: Bool) {
+        var components: [Substring] = self.split(path)
+        
+        let isAbsolute: Bool
+        if path.first == PathComponentScanner.separator {
+            guard let maybeModuleName = components.first?.dropFirst(), !maybeModuleName.isEmpty else {
+                return ([], true)
```

Comment:
When would this line be reached? for a path that begins with `//` ?

Reply:
This is reached if the path just a single `/`. If that path was `//` there would still be a non-empty string after dropping the leading slash.